descqa/NumberDensityVersusRedshift.py

```diff
@@ -0,0 +1,307 @@
+            #initialize default values to use for validation-data fits
+            z0 = 0.
+            z0err = 0.
+            #get masks for magnitude cuts and get validation-data fits depending on selected observation
+            if self.observation.find(Coil2004)!= -1:
+                mask = (catalog_data[mag_field]<=cut) & (catalog_data[mag_field]>self.mag_hi[n])
+                cutlabel = ' '.join(['$',str(self.mag_hi[n]),'< $',self.band,'$<=',str(cut),'$'])
+                
+                #find correct row in validation data fits to use in plot
+                omask = (self.mag_hi.astype(int)==int(self.mag_hi[n])) & (self.mag_lo.astype(int)==int(cut))
+                z0 = self.validation_data['z0values'][omask]
+                if('z0errors' in self.validation_data.keys()):
+                    z0err = self.validation_data['z0errors'][omask]
+            else:
+                mask = (catalog_data[mag_field]<cut)
+                cutlabel =  ' '.join([self.band,'$<',str(cut),'$'])
+                if(self.observation.find(DEEP2_JAN)!= -1):
+                    z0 = self.validation_data['z0const'] + self.validation_data['z0linear']*(cut)
+            total = '(# of galaxies = {})'.format(np.sum(mask))
+ 
+            #bin catalog_data
+            N,binEdges = np.histogram(catalog_data[self.z][mask],bins=self.zbins)
+            sumz,binEdges = np.histogram(catalog_data[self.z][mask],bins=self.zbins,weights=catalog_data[self.z][mask])
+            meanz = sumz/N
+
+            #make subplot
+            ncol = int(n%self.ncolumns)
+            nrow = n//self.ncolumns
+            catalog_label = ' '.join([catalog_name,re.sub(self.band,filtername+' '+self.band,cutlabel)])
+            validation_label = ' '.join([self.validation_data['label'],cutlabel])
+            reskey = re.sub('\$','',cutlabel)
+            results[reskey] = self.make_subplot(meanz,catalog_data[self.z][mask],n,ax[nrow,ncol],z0,z0err,catalog_color,validation_color,catalog_label,validation_label)
+            results[reskey]['total'] = total
+
+            #add curve for this catalog to summary plot
+            if self.first_pass:  #add validation data if evaluating first catalog
+                summary = self.make_subplot(meanz,catalog_data[self.z][mask],n,self.summary_ax[nrow,ncol],z0,z0err,catalog_color,validation_color,catalog_label,validation_label)
+            else:
+                summary = self.make_subplot(meanz,catalog_data[self.z][mask],n,self.summary_ax[nrow,ncol],0.,0.,catalog_color,validation_color,catalog_label,validation_label)
+
+        #make empty subplots invisible 
+        if len(self.mag_lo)%self.ncolumns!=0: #check for empty subplots
+                ax[self.nrows-1,self.ncolumns-1].set_visible(False)   #assumes ncolumns=2
+                
+        #save results for catalog and validation data in txt files
+        for filename, dtype, comment, info in zip([filelabel,self.observation],['y','fit'],[filtername,''],['total','']):
```

Comment:
`for filename, dtype, comment, info in zip_longest((filelabel, self.observation), ('y', 'fit'), (filtername,), ('total',)):`

Reply:
Done